### ISSUE TITLE
[CT-2370] Inconsistent behaviour of dbt seed and column quoting for Postgres and Snowflake

### ISSUE THREAD

**Issue by az-hienas** [dbt-labs/dbt-core#7265]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

**Initial note**
The issue is related to inconsistent behaviour of `dbt seed` between **dbt-postgres** and **dbt-snowflake**.
I wasn't sure where to post the issue, so I chose to post in in the **dbt-core** repo. Please let me know if I should post this issue somewhere else.

The issue is about the difference in behaviour of `dbt seed` when creating columns that might have different cases. In Postgres the columns are quoted if the column names are not all lowercase. In Snwoflake the columns are not quoted even if the CSV header contains columns of different cases.

Furthermore, if a column name in the seed CSV starts with a number Postgres will automatically quote it, whereas in Snowflake will return an error (although Snowflake allows for column names to start with a number as long as they are quoted).

### Expected Behavior

Consistent behaviour between Postgres and Snowflake. Either quote the columns and keep the original case from the CSV or not quote at all. 
Also the Postgres column quoting if there is a difference in case in the CSV headers seems to go against the documentation at: 
https://docs.getdbt.com/reference/resource-configs/quote_columns
which specifies that the default behaviour is to not quote:
> - (Default) When False, dbt will not quote the column names defined in the seed file.

### Steps To Reproduce

**Example 1: regarding different cases**
Given the CSV file (e.g. named `people.csv`:
```
Name,Age
Sarah,25
John,30
Nick,18
Mandy,4
```

Running `dbt seed` on the above CSV in Postgres will quote the columns so that they retain the case and the columns can only be accessed if quoted:

``` sql
-- people table with data from people.csv
select  "Name", "Age"
from people
```

Not quoting the `Name` and `Age` column in the Postgres `SELECT` will result in an error. 

Running `dbt seed` on the above CSV in Snowflake will not quote the columns and as a result any combination of upper and lowercase letters of the `Name` and `Age` column will work:

``` sql
select NaMe, aGe
from people;
``` 

However if the columns are quoted (i.e. `select "Name", "Age" from people`) Snowflake will return an error.

**Example 2: regarding columns starting with a number**
Given the CSV file (second version of `people.csv`)
```
1stname,age
Sarah,25
John,30
Nick,18
Mandy,4
```
Running `dbt seed` on the above CSV  in Postgres will quote the `1stname` column and not quote the `age` column.

``` sql
-- people table with data from second version of people.csv
select "1stname", age
from people;
```

Running `dbt seed` on the above CSV in Snowflake will result in an error:
```
Database Error in seed people (fake_seeds/fake_test_seeds/people.csv)
  001003 (42000): SQL compilation error:
  syntax error line 1 at position 40 unexpected '1'
```

### Environment

```markdown
- OS: Debian GNU/Linux 11 (bullseye) - Docker image (FROM --platform=linux/amd64 mcr.microsoft.com/devcontainers/python:0-3.9-bullseye)
- Python: 3.9.16
- dbt-core: 1.3.0
- dbt-postgres: 1.3.0
- dbt-snowflake: 1.3.0
```

**Comment by dbeatty10** [comment]
Thanks for reporting your experience @az-hienas !

### TLDR
The default quoting of columns differing between Snowflake and Postgres is expected actually! (See "More detail" below for the explanation.)

But you can fully control the behavior by explicitly [quoting seed columns globally](https://docs.getdbt.com/reference/resource-configs/quote_columns#globally-quote-all-seed-columns) in `dbt_project.yml`:
```yaml
seeds:
  +quote_columns: true
```

Use `quote_columns: false` to do the opposite.

### More detail

By default, the `quote_columns` configuration for seed columns is [`False` for Snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/aa0401129b674e6126e841a82e5de899b5c18470/dbt/adapters/snowflake/impl.py#L144-L145) and [`True` for Postgres](https://github.com/dbt-labs/dbt-core/blob/ae485f996a4a92edfea7fe9f3d0cf9f9d3a255e6/core/dbt/adapters/base/impl.py#L836-L838
)

This matches the `quoting` configuration that applies to database, schema, and table/view names which is set to `false` [by default](https://docs.getdbt.com/reference/project-configs/quoting#default) for Snowflake and `true` for Postgres.

You can read more about the rationale for these differences [here](https://docs.getdbt.com/reference/project-configs/quoting#snowflake):
> Whereas most databases will lowercase unquoted identifiers, Snowflake will uppercase unquoted identifiers. If a model name is lowercased and quoted, then it cannot be referred to without quotes!

### Summary
Since the default is behaving as expected and you can explicitly configure different behavior, I'm going to close this as `wont_fix`. We can reconsider re-opening if there's some key detail I missed.

**Comment by dbeatty10** [comment]
I forgot to mention that I agree with your reflection here:

> Also the Postgres column quoting if there is a difference in case in the CSV headers seems to go against the documentation at:
> https://docs.getdbt.com/reference/resource-configs/quote_columns
> which specifies that the default behaviour is to not quote:
> 
> > - (Default) When False, dbt will not quote the column names defined in the seed file.

We should update the documentation to be something more like the following:

- When `True`, dbt will quote the column names defined in the seed file when building a table for the seed, preserving casing.
- When `False`, dbt will not quote the column names defined in the seed file.
- When not set, the default will vary by adapter.

I think we can remove this portion altogether in favor of the [recommended configuration](https://docs.getdbt.com/reference/resource-configs/quote_columns#recommended-configuration) section:
> Changelog
> - Future: The default value may change in a future release. If you're using seed files, it is recommended that you set this configuration explicitly to avoid breaking changes in the future.

**Comment by az-hienas** [comment]
@dbeatty10 Thanks for the reply, I will just make it for now so that the column names are all lowercase and start with a letter, that way there shouldn't be any problems.
